### ISSUE TITLE
Work flow for contributors

### ISSUE THREAD

**Issue by santoshphilip** [santoshphilip/eppy#86]
If you are a contributor, consider using the following workflow
- an issue is opened
- the issue is assigned to you
- You make a branch from `develop` with a meaningful name such as `i57_licence` where `i57` is the issue number
- Once the coding is complete and the tests pass, inform the "release master"  of the repository.( release master   = @santoshphilip at present)
- "release master" will review and merge your changes into `develop`
- close the issue

If you are not a contributor
- ask to become a contributor

OR

- make a pull request
- make changes in the develop branch
- request a merge into the repository

(not sure I used the terminology correctly in the last few steps :-(

**Comment by jamiebull1** [comment]
@santoshphilip While I agree with most of this, I quite like the pull request approach since it avoids clutter in the main repository. As it stands, there are quite a few unfinished branches from various contributors that new contributors end up downloading. If they had been pull requests from other repos then they wouldn't be fetched when you `git clone` this repo. It's probably not a big deal with a relatively small number of contributors, but as time goes by the load will tend to increase.

Also, I'd rephrase the instructions for this route as:

- fork this repository into your own repository
- make a branch off the `develop` branch to work on your new feature
- once the coding is complete, raise a pull request against the `develop` branch here
- give the pull request a meaningful name, and mention any issues that it addresses
- once the code has been reviewed and tests are passing it will be merged into `develop`

**Comment by santoshphilip** [comment]
if you are in my shoes you would not like the pull request strategy at all
Pull request is very opaque from here.
Until the merge request comes in I am flying blind.

I the updates in a branch, without the pull request, I can see the push happen.
So I know there is some activity going on (usually I don't look too deep into it)
So I know if I have to check with someone, reassign the task or just let it be.

I am going off @digwanderlust suggestion here.
- core contributors make a branch and work on it.
- others may do a pull request and work on it and ask for a merge. This is usually one time contributions

With more pull requests coming in, I can see the rationale behind it.

Regarding the many branches hanging around. That is a result of me being a packrat. Almost all of them should be deleted. As far I can make out, the standard practice is to delete a branch once it has been merges to a main branch (develop or master). I'll open an issue to delete the finished branches.

**Comment by lafrech** [comment]
Notes:

- You may use `git branch -d iXX_dev_branch` to delete a branch. It will delete it only if it was merged. Doing this on all dev branches should only remove those that can be. Removing an unmerged branch takes a `-d --force` or equivalently `-D`. ([git-branch](https://git-scm.com/docs/git-branch))

- You may want to put the content of this issue in a `CONTRIBUTE` file so that it is displayed to anyone starting to contribute (like when creating an issue, for instance). (See GitHub doc [here](https://github.com/blog/1184-contributing-guidelines) and [here](https://help.github.com/articles/setting-guidelines-for-repository-contributors/).)

**Comment by jamiebull1** [comment]
+1 for a `CONTRIBUTE` file

**Comment by jamiebull1** [comment]
Here's my suggested text for `CONTRIBUTE.md`

If you are a contributor, consider using the following workflow:

-     an issue is opened
-     the issue is assigned to you
-     you make a branch from `develop` with a meaningful name such as "i57_licence" where #57 is the issue number
-     once the coding is complete and the tests pass, raise a pull request against the `develop` branch and tag the "release master" of the repository (release master is @santoshphilip at present)
-     "release master" will review and merge your changes into the `develop` branch
-     close the issue

If you are not a contributor:

-     ask to become a contributor

OR

-     fork this repository into your own repository
-     make a branch off the `develop` branch to work on your new feature
-     once the coding is complete, raise a pull request against the `develop` branch here
-     give the pull request a meaningful name, and mention any issues that it addresses (using the issue number, e.g. "Resolves #86, also see #57")
-     once the code has been reviewed and tests are passing it will be merged into `develop`

**Comment by santoshphilip** [comment]
Looks good. 
Will start to incorporate this.

**Comment by jamiebull1** [comment]
Closing. Discussion can continue on PR #133
